### ISSUE TITLE
MQTT is being disconnected when device is not registered but gateway isn't. Expecting an error message on topic instead of disconnect

### ISSUE THREAD

**Issue by BobClaerhout** [eclipse/hono#3039]
We have an authenticated gateway sending mqtt messages on behalf of devices to t/{tenant}/{gatewayId}/mySubject. The gateway is also subscribed to the error topic to avoid having those disconnects. The gateway has a downstream mapper configured and that mapper returns a different deviceId. When this device is not registered (or disabled) we are experiencing following:

- we receive an exception on the error topic stating that the device does not exist
- we are being disconnected.

The disconnect is a bit strange since the gateway is registered properly and we are able to use this gateway to send messages for other devices.
I think following piece of code (found in `AbstractProtocolAdapterBase.isTerminalError`) is the reason why this disconnect happens, it seems to be identified as a terminal error even while the gateway is properly registered.
```
// if the device is not registered or disabled
if (error instanceof DeviceDisabledOrNotRegisteredException) {
    // and if the device is connected via a gateway
    if (deviceId != null && !authenticatedDevice.getDeviceId().equals(deviceId)) {
        return getRegistrationAssertion(authenticatedDevice.getTenantId(),
                authenticatedDevice.getDeviceId(), null, spanContext)
                        .map(ok -> false)
                        .recover(e -> {
                            // and if the gateway is not registered then it is a terminal error
                            return Future
                                    .succeededFuture(e instanceof DeviceDisabledOrNotRegisteredException);
                        });
    } else {
        return Future.succeededFuture(true);
    }
}
```
This might be due to the fact that the mapper is actually tampering the incoming message (and might be changing the authenticatedDeviceId but not sure of this).

**Comment by calohmn** [comment]
When the downstream mapper invoked during the processing of an MQTT message returns a mapped message with a different device id, that device id is not used when invoking the `AbstractProtocolAdapterBase.isTerminalError` method. Namely, the `MqttContext` object isn't updated with the device id retrieved from the mapper.
Since the original `MqttContext` device id is the one from the gateway, the `DeviceDisabledOrNotRegisteredException` is interpreted to refer to the gateway not being enabled or registered. Therefore the gateway gets disconnected.

**Comment by sophokles73** [comment]
closed via #3040
